### ISSUE TITLE
charging.launch does not work

### ISSUE THREAD

**Issue by bfalacerda** [strands-project/scitos_apps#37]
The charging.launch file does not work, due to this line:
    `<remap from="/cmd_vel" to="/cmd_vel_mux/input/navigation"/>`

From what I understand this cmd_vel_mux topic is used so that we can have different priorities in terms of what is published to cmd_vel (e.g., rumblepad > move_base). However, I'm not sure where this remapping needs to be done, and whether it's worth using this at all. Is is worth the hassle?

**Comment by gestom** [comment]
The main advantage of the mux is that it allows to take full manual control of the robot in case some of the components fail and send nonsensical commands to the /cmd_vel. Without the mux, the joystick commands will mix with the wrong ones resulting in erratic robot behaviour. 

IMHO, that is quite essential.

**Comment by cburbridge** [comment]
This means that some program might be controlling the robot, but being blocked by the command multiplexer. How does a program know whether its velocity commands will make it to the robot or not? 

I can see that if the program fails badly and tries to drive the robot dangerously, you want to grab the joystick and block it. But you can do this with the motor_stop service. Granted, by multiplexing the cmd_vel you can not just stop the robot but take control, but is the added level of complexity incurred really worth it? 

I think it would be better to detect an override signal (eg joystick button) from within the controlling program, and stop sending /cmd_vel.

**Comment by gestom** [comment]
at §1) If a control program needs to know if its commands are making it to the robot, then comparing the /odom feedback to the send commands is one of the possible ways of doing that. Still, it seems to me that such a case is quite rare. Could you give a real-world example when it's worth to rely on /cmd_vel rather than on /odometry feedback ?

ad §2) Does the motor_stop still allow to drive the robot around by joystick ? 
IMHO, the ability to drive the robot around if all the high-level stuff fails is worth one extra node in a system, which has more that 30 nodes already and will be running more than 50 nodes.

ad §3) Overriding a /cmd_vel signal from within each control program means that every control program includes the override routine, which is always working. 

Also, at some point, we will have to solve action selection problem. Either we will have to assume that all the nodes, which might send commands to the /cmd_vel are always behaving properly, i.e. they can be paused or stopped and resumed on request, or we have one extra node, which ensures that even if the high-level control modules fail, the more "critical" ones will override the commands from the less critical ones. After all, having the mux allows us to implement some nice ideas of the subsumption architecture-based robots, which are quite successful at longterm autonomy already.

**Comment by cburbridge** [comment]
The way I see the system working is to have an overall state machine that schedules task on the robot. So far, we have a 'scheduler' that simply selects a waypoint and drives there, or chooses to charge the robot. In the future we  will have additional tasks, such as monitoring desks, looking for objects, finding a person, allowing teleop, etc. Tasks should always be preemptable - the scheduler would be then able to stop any control program from trying to control the robot.

So, §1: suppose the scheduler has started a task that moves the robot to a waypoint. The cmd_vel messages are coming from the navigation stack, so the scheduler now has to subscribe to the cmd_vel from the current task in addition to the odometry, and continuously monitor the difference. Now if the robot is stuck on the carpet, or in collision with an object, then odometry does not reflect the cmd_vel. Does this mean that the scheduler should assume that a user has taken control with the joystick? Or also subscribe to the bumper state? 

§2: motor stop does not allow the robot to move, but if the scheduler has failed to the point of not preempting tasks, then it would probably need restarting, being able to manually move the robot won't help much. Running one extra node is obviously not much in a large system, but the change in configuration of components, and in particular to the logic of the scheduler is what concerns me.

§3: I think each tasks/program would have to be wrapped as an actionlib action or similar to be invoked by the scheduler. So then it also has to be preemptable by the scheduler. If preempting fails, then the scheduler can further try to kill the task, but either way, leaving it running with cmd_vel publishing but not doing anything seems wrong.

So, overall I don't see that the multiplexing gains anything extra that could not be done more flexibly with WP6's scheduler.

**Comment by gestom** [comment]
I totally agree. All tasks should be preemtable. But we simply cannot assume perfect implementation of all the nodes publishing to /cmd_vel. The mux simply allows to prioritize the commands in case any of the nodes fail to be preempted. Moreover, the scheduler itself might instruct the mux to block the commands from preemted modules. Thus, the assumption of having all node implementation to be preemptable at any moment is substituted by having one node, i.e. the mux implemented correctly. Don't get me wrong, I agree that the tasks should be preeemptable, but the mux is just an additional safety feature in case some of the tasks cannot be stopped. IMHO, considering that the software will continuously evolve and never be fixed, assumption that all nodes don't contain bugs that might prevent their preemtion, is unrealistic.

§1) Yeap, stuck detection (e.g. on a carpet) is a good example. However stuck detection should not be part a scheduler, but a separate 'low-level' process, because it basically acts as a virtual sensor. Having the mux part of the system actually does not mean that you do not have access to /cmd_vel any more, so the carpet detection can still be done in the same way as before. 

§2) As far as I understand, the only change is remapping the /cmd_vel to /cmd_vel_mux for the processes, which publish to /cmd_vel.  

§3) I agree. But it probably takes some time to identify and kill a fautly node. As I have mentioned before, in addition to starting and pausing the individual processes, the scheduler can also instruct the mux to accept input only from the nodes which are supposed to be active, which prevents the faulty nodes from causing erratic behaviour of the robot.

**Comment by bfalacerda** [comment]
I think that if we have a faulty non-preemptable node causing erratic behaviour in the robot, we don't really need the ability to teleop the robot around: we are pretty much in an unrecoverable situation, so it is a fatal fault and the run should be considered a failure. Thus, in my opinion, being able to stop the robot by pressing a button is enough.

The situations where I think being able to drive the robot around is important are, for example, situations where the robot is navigating to dangerous areas or is stuck (this sometimes happens when running move_base). In such a case, a simple "gamepad button pressed" monitor in the state machine can be used to jump to a state where nothing is being sent to /cmd_vel and the teleoperation can be used. The monitor and recovery states work very well in these cases, as is proved by the stuck on carpet monitor and recover states. In my opinion, it is better for the controller state machine to have full control and knowledge of what is published where, without an extra node changing its commands. This structure also has the advantage of easily providing the possibility of logging the fact that teleoperation had to be used.

I'd also like to make a "best practices" comment related to this: When something is added that requires other components of our system to adhere to a certain scheme (in this case, all nodes publishing to /cmd_vel need to do a remapping), it cannot be expected that they will be immediately changed to adhere to the new rules. I think it's up to the people that add something like this, and know exactly what it is doing, to change the remaining components so that everything remains correctly integrated. In this case, if a pull request in scitos_2d_navigation adding the remapping and the launching of the mux node had been done when this package was added, we could have had this discussion then. Furthermore, it would have avoided having other members of the consortium finding out on their own why the charging.launch file is not working in the long_term_patroller.

**Comment by gestom** [comment]
ad §1) According to @cburbridge comment 2.3, the faulty nodes can be automatically identified and terminated.
In such a case, the mux will prevent the faulty node to do too much harm before being identified and killed.

ad §2) Yeap, this approach works perfectly if all nodes implemeted correctly.

ad §3) Sorry, my mistake, I apologise for that. The mux has been created just after the RoAR and I did not realize that its not part of the complete system everyone is running. However, it's purely optional and whoever does not want to use it, does not have to. It just adds a functionality that I find useful. I will modify the "charging.launch" to reflect that "best practices" are more important than overall system robustness.

**Comment by bfalacerda** [comment]
My point was not changing the docking.launch file, we are still discussing on wether it's worth using the mux stuff or not. It is clearly a possible solution, just not the one I would prefer, but if more people prefer it, I'll be happy to go for it.

I just wanted to point out that one should make sure that other parts of the system that are affected by a new functionality are altered to reflect that. In this case, scitos_2d_navigation wasn't, and people continued using it as it was, while this mux approach was a bit "hidden" in this repo.

**Comment by hawesie** [comment]
Yes, agreed that we shouldn't introduce changes that affect the rest of the system without proper consultation. We should also be more alert for things like this when pull requests come in. This is the time when we should be checking for issues that affect us.

**Comment by marc-hanheide** [comment]
Agreed, it was unfortunate to make this change having such a big impact. It's not quite clear to me yet, how we will approve changes that effect several repositories at the same time. There is no "atomic action" We'd probably have to have work more with "stable tags" to define a working system configuration...

But I want to stress that arbitration/action selection is indeed a question that needs to be resolved. This is true for base movements as well as (and probably even more often) the PTU and head movements. I can already see that the obstacle avoidance want to have the pan-tilt pointed forward and tilted down to see obstacles, while at the same time the person detection want the PTU look straight or even sweeping. How are we going to accommodate these different requirements stemming from concurrent processes access actuators of the robot?

The mux strategy is a first proposal on this, which generally has the advantage of being not too invasive (components can still publish topics as they used to), while still being rather flexible. To make it more generally applicable, we can allow a dynamic reconfiguration of that muxer, e.g. from planning/scheduling/smach. For that, I'd suggest that the muxer implements a service/actionlib interface or simply updates the priorities from the param server, so that it can be externally configured to allow flexible arbitration. 

It requires more general discussion. There are, AFAIK, only very few cmd-topics that will be "mux-ed", namely the ones for vel, ptu, and head. Maybe we should include them in general to allow for the flexibility it provides, as I cannot see any negative impacts once it is established across the system. I might have overlooked something, though.

Final question: @hawesie, what is the best place within STRANDS to have these more general discussions? Somewhere hidden in an issue in one of the repositories probably isn't the right place.

**Comment by hawesie** [comment]
Can we discuss this at the marathon Skype next week? I'd say this level of detail should either be on the strands_management issue board (as it's private and not tied to specific code) or on the strands-wp8 mailing list. 

@marc-hanheide shouldn't the patrolling Jenkins test have picked this up?

**Comment by marc-hanheide** [comment]
Yes, it should. But it's not working ATM for other reasons (i.e. not using the data centre correctly, because there was an issue of not being able to run it on another database bath that just got resolved), so didn't test it. I'll update the test as soon as I can, i.e. not today ;-)

**Comment by hawesie** [comment]
Cool :)
